futures-core/src/future.rs

```diff
@@ -5,6 +5,19 @@ use core::pin::PinMut;
 
 pub use core::future::{Future, FutureObj, LocalFutureObj, UnsafeFutureObj};
 
+/// A `Future` or `TryFuture` which tracks whether or not the underlying future
+/// should no longer be polled.
+///
+/// `can_poll` will return `false` if a future should no longer be polled.
+/// Usually, this state occurs after `poll` (or `try_poll`) returned
+/// `Poll::Ready`. However, `can_poll` may also return `false` if a future
+/// has become inactive and can no longer make progress and should be ignored
+/// or dropped rather than being `poll`ed again.
+pub trait FusedFuture {
+    /// Returns `false` if the underlying future should no longer be polled.
+    fn can_poll(&self) -> bool;
```

Comment:
I do not think this is a good name as it is not clear that it implies the future being complete. `can_poll` initially read to me as a hint for being ready.

`is_complete()`, `is_final`, ... would be much better.

Reply:
I tried so many darn names for this thing! I had `has_completed()` for a while, but IMO it's misleading because it sounds like a "complete" future would be one that is done working and ready to return a value, and for streams it's even more unclear what "complete" means, because it's not the point at which it has yielded every element-- it's the point at which the stream has already yielded `None`.

There's also the interesting case that I was trying to be compatible with where we want to support `FuturesUnordered` having a mode where `can_poll` returns false when there are no elements left, rather than repeatedly resolving to `None` (which is often an uninteresting case). e.g.:

```rust
let mut running_tasks = FuturesUnordered::new();
running_tasks.no_none_yield(); // bikeshed away
loop {
    let mut recv = rx.next().fuse();
    select! {
        future(recv as msg) => {
            running_tasks.push(async { ... })
        }
        next(running_tasks as completed_task) => { ... }
        all complete => { return ... }
    }
}
```

It may be we make a thin wrapper around `FuturesUnordered` called e.g. `TaskSet` or something for this, but the point is that for whatever type this is, `can_poll` refers to something different from "has yielded `None` already".

Reply:
`is_terminated`?

Reply:
Sure, I'd be fine with `is_terminated`, where "terminated" is something like "has reached a state where polling would yield redundant or inconsistent results, or otherwise misbehave". Does that sound about right to you?